File: internal/model/client-user.go
package model

import "github.com/usesnipet/snipet/pkg/jsonx"

type ClientUser struct {
	ID string `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`

	Name     string        `gorm:"type:varchar(255);not null" json:"name"`
	Picture  *string       `gorm:"type:text" json:"picture"`
	Email    *string       `gorm:"type:text" json:"email"`
	Metadata jsonx.JSONMap `gorm:"type:jsonb;not null" json:"metadata"`

	ClientUserToSessions []ClientUserToSession `gorm:"foreignKey:ClientUserID;references:ID;constraint:OnDelete:CASCADE" json:"client_user_to_sessions"`
	ClientToClientUsers  []ClientToClientUser  `gorm:"foreignKey:ClientUserID;references:ID;constraint:OnDelete:CASCADE" json:"client_to_client_users"`
}

type ClientToClientUser struct {
	ClientID     string  `gorm:"primaryKey" json:"client_id"`
	ClientUserID string  `gorm:"primaryKey" json:"client_user_id"`
	ExternalID   *string `gorm:"type:varchar(255);index" json:"external_id"`

	Client     *Client     `gorm:"foreignKey:ClientID;references:ID;constraint:OnDelete:CASCADE" json:"client"`
	ClientUser *ClientUser `gorm:"foreignKey:ClientUserID;references:ID;constraint:OnDelete:CASCADE" json:"client_user"`
}

type ClientUserToSession struct {
	ClientUserID string `gorm:"primaryKey" json:"user_id"`
	SessionID    string `gorm:"primaryKey" json:"session_id"`

	ClientUser *ClientUser `gorm:"foreignKey:ClientUserID;references:ID;constraint:OnDelete:CASCADE" json:"client_user"`
	Session    *Session    `gorm:"foreignKey:SessionID;references:ID;constraint:OnDelete:CASCADE" json:"session"`
}

File: web/src/models/client-user.ts
import { z } from "zod";

import { clientSchema, type Client } from "@/models/client";
import { sessionSchema, type Session } from "@/models/session";

export const clientUserMetadataSchema = z.record(z.string(), z.unknown());
export type ClientUserMetadata = z.infer<typeof clientUserMetadataSchema>;

export interface ClientUser {
  id: string;
  name: string;
  picture?: string | null;
  email?: string | null;
  metadata: ClientUserMetadata;
  client_user_to_sessions: ClientUserToSession[] | null;
  client_to_client_users: ClientToClientUser[] | null;
}

export const clientUserSchema: z.ZodType<ClientUser> = z.lazy(() =>
  z
    .object({
      id: z.uuid(),
      name: z.string(),
      picture: z.string().nullable().optional(),
      email: z.string().nullable().optional(),
      metadata: clientUserMetadataSchema,
      client_user_to_sessions: z.array(clientUserToSessionSchema).nullable(),
      client_to_client_users: z.array(clientToClientUserSchema).nullable(),
    })
    .strict(),
);

export interface ClientToClientUser {
  client_id: string;
  client_user_id: string;
  external_id: string | null;
  client?: Client | null;
  client_user?: ClientUser | null;
}

export const clientToClientUserSchema: z.ZodType<ClientToClientUser> = z.lazy(() =>
  z
    .object({
      client_id: z.uuid(),
      client_user_id: z.uuid(),
      external_id: z.string().nullable(),
      client: clientSchema.nullable().optional(),
      client_user: clientUserSchema.nullable().optional(),
    })
    .strict(),
);

export interface ClientUserToSession {
  user_id: string;
  session_id: string;
  client_user?: ClientUser | null;
  session?: Session | null;
}

export const clientUserToSessionSchema: z.ZodType<ClientUserToSession> = z.lazy(() =>
  z
    .object({
      user_id: z.uuid(),
      session_id: z.uuid(),
      client_user: clientUserSchema.nullable().optional(),
      session: sessionSchema.nullable().optional(),
    })
    .strict(),
);
